capsules/src/virtual_uart.rs

```diff
@@ -0,0 +1,217 @@
+        if self.receive_in_progress.get() == false {
+            self.rx_buffer.take().map(|rxbuf| {
+                self.receive_in_progress.set(true);
+                let len = cmp::min(rx_len, rxbuf.len());
+                self.uart.receive(rxbuf, len);
+            });
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Operation {
+    Transmit { len: usize },
+}
+
+pub struct UartDevice<'a> {
+    mux: &'a UartMux<'a>,
+    receiver: bool, // Whether or not to pass this UartDevice incoming messages.
+    tx_buffer: TakeCell<'static, [u8]>,
+    rx_buffer: TakeCell<'static, [u8]>,
+    rx_len: Cell<usize>,
+    operation: OptionalCell<Operation>,
+    next: ListLink<'a, UartDevice<'a>>,
+    client: OptionalCell<&'a hil::uart::Client>,
+}
+
+impl<'a> UartDevice<'a> {
+    pub const fn new(mux: &'a UartMux<'a>, receiver: bool) -> UartDevice<'a> {
+        UartDevice {
+            mux: mux,
+            receiver: receiver,
+            tx_buffer: TakeCell::empty(),
+            rx_buffer: TakeCell::empty(),
+            rx_len: Cell::new(0),
+            operation: OptionalCell::empty(),
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    /// Must be called right after `static_init!()`.
+    pub fn setup(&'a self) {
+        self.mux.devices.push_head(self);
+    }
+}
+
+impl<'a> hil::uart::Client for UartDevice<'a> {
+    fn transmit_complete(&self, tx_buffer: &'static mut [u8], error: hil::uart::Error) {
+        self.client.map(move |client| {
+            client.transmit_complete(tx_buffer, error);
+        });
+    }
+
+    fn receive_complete(
+        &self,
+        rx_buffer: &'static mut [u8],
+        rx_len: usize,
+        error: hil::uart::Error,
+    ) {
+        self.client.map(move |client| {
+            client.receive_complete(rx_buffer, rx_len, error);
+        });
+    }
+}
+
+impl<'a> ListNode<'a, UartDevice<'a>> for UartDevice<'a> {
+    fn next(&'a self) -> &'a ListLink<'a, UartDevice<'a>> {
+        &self.next
+    }
+}
+
+impl<'a> hil::uart::UART for UartDevice<'a> {
+    fn set_client(&self, client: &'a hil::uart::Client) {
+        self.client.set(client);
+    }
+
+    // Ideally this wouldn't be here, and if we ever create a "UartDevice" trait
+    // in the HIL then this could be removed.
+    fn init(&self, params: hil::uart::UARTParams) {
+        self.mux.uart.init(params);
+    }
+
+    /// Transmit data.
+    fn transmit(&self, tx_data: &'static mut [u8], tx_len: usize) {
+        self.tx_buffer.replace(tx_data);
+        self.operation.set(Operation::Transmit { len: tx_len });
+        self.mux.do_next_op();
+    }
+
+    /// Receive data until buffer is full.
+    fn receive(&self, rx_buffer: &'static mut [u8], rx_len: usize) {
+        self.rx_buffer.replace(rx_buffer);
+        self.rx_len.set(rx_len);
+        self.mux.start_receive(rx_len);
+    }
+
+    // This is not virtualized. Calling this will cause all receives to stop.
```

Comment:
I think this could virtualize fairly easily by changing this to set the `receiver` flags for this instance to false, then iterating all the mux'd devices and actually calling `abort` only if all devices are set not to receive

Reply:
That would be the opposite nonvirtualized behavior, correct? Right now it is first to call abort aborts for all, and that would be last to call abort blocks early receive for all.

Reply:
Hmm, I hadn't realized that aborting returned partial buffers. That's a bit unfortunate. As I don't think much (anything) is built on fixed buffer receive + abort yet, maybe we should have a discussion around this HIL and ABI.

Reply:
I think I'd prefer this being a no-op (or even an `unimplemented!`) in the short term, to discourage anything from trying to use it, as the current behavior can never virtualize and it will likely go away in eventual UART HIL refinements.